config/tools/config_generator.cpp

```diff
@@ -0,0 +1,935 @@
+    }
+
+    [[maybe_unused]] auto config_generator::generate_configuration_file()
+        -> std::string {
+        // Cumulative return message
+        std::string return_msg = "";
+        std::string output_filename = "tmp.cfg";
+
+        if(!template_file_is_valid) {
+            std::filesystem::path temp_build_dir = m_build_dir;
+            temp_build_dir.append("config").append("tools");
+            return_msg += "File provided, " + m_template_config_file
+                        + ", did not exist and could not be copied to "
+                        + temp_build_dir.string()
+                        + ". Aborting operation. Please rerun with proper "
+                          "template location \n";
+            return return_msg;
+        }
+
+        std::map<std::string, std::string> config_map;
+        load_template(m_template_config_file, config_map);
+        // Deal with all config values in file with "tmpl_"
+        for(auto config_map_it = config_map.begin();
+            config_map_it != config_map.end();
+            config_map_it++) {
+            if(config_map_it->first.find(template_prefix)
+               != std::string::npos) {
+                template_options.emplace(
+                    config_map_it->first,
+                    parse_value(config_map_it->second, false));
+
+            } else {
+                value_t parsed_val = parse_value(config_map_it->second, true);
+                if(std::holds_alternative<size_t>(parsed_val)) {
+                    set_param_to_config_file(config_map_it->first,
+                                             std::get<size_t>(parsed_val));
+                } else if(std::holds_alternative<double>(parsed_val)) {
+                    set_param_to_config_file(config_map_it->first,
+                                             std::get<double>(parsed_val));
+                } else if(std::holds_alternative<std::string>(parsed_val)) {
+                    set_param_to_config_file(
+                        config_map_it->first,
+                        std::get<std::string>(parsed_val));
+                } else {
+                    __builtin_unreachable();
+                }
+            }
+        }
+
+        // Create Components
+        const auto is_two_phase_mode
+            = get_param_from_template_file(two_phase_mode, config_map);
+        const auto shard_count
+            = get_param_from_template_file(shard_count_key, config_map);
+        const auto sentinel_count
+            = get_param_from_template_file(sentinel_count_key, config_map);
+
+        if(std::holds_alternative<size_t>(is_two_phase_mode)
+           && std::get<size_t>(is_two_phase_mode) == 1) {
+            output_filename = "2pc_generated_config.cfg";
+            const auto coordinator_count
+                = get_param_from_template_file(coordinator_count_key,
+                                               config_map);
+            if(std::holds_alternative<size_t>(sentinel_count)) {
```

Comment:
This is another process that's repeated a lot. Does it seem reasonable to refactor this (if only for maintainability)?

Reply:
DONE